### PR TITLE
C#: Improve db consistency by removing assembly id

### PR DESCRIPTION
**Pull request by calumgrant**
This addresses an issue that occurs particularly with .NET Core, where classes can be defined in more than one assembly. Particularly, the assembly `Private.CoreLib` duplicates much of the standard library, from `System.Runtime` and so on. Relaxing the qualifiers on IDs in some places removes many of the dangling references in the database.

### CODE REVIEW

File: csharp/ql/test/library-tests/aliases/aliases2.expected [hunk 1]
```diff
@@ -1,3 +1,9 @@
 | Program.cs:18:21:18:22 | c1 | Assembly1.dll:0:0:0:0 | Class |
+| Program.cs:18:21:18:22 | c1 | Assembly2.dll:0:0:0:0 | Class |
```
Comment (calumgrant) [inline csharp/ql/test/library-tests/aliases/aliases2.expected]: @hvitved I'm not sure whether this is a desired change in behaviour. The reason for the change is that the local variable has a type given by a `typeref` which is no longer qualified with the assembly id. Therefore, the type of the local variable maps to all matching types, which is three different types in this case.

To recover the original behaviour, the extractor simply needs to populate a `type` as opposed to a `typeref` in the local variable's `type` field.

Comment (hvitved) [reply csharp/ql/test/library-tests/aliases/aliases2.expected]: I don't think types of local variables should be treated specially, what do you think?

File: csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs [hunk 2]
```diff
@@ -130,10 +148,12 @@ public static bool ContainsTypeParameters(this ITypeSymbol type, Context cx, ISy
...
         /// <param name="cx">The extraction context.</param>
         /// <param name="trapFile">The trap builder used to store the result.</param>
+        /// <param name="prefix">Whether to prefix the type ID with the assembly ID.</param>
+        /// <param name="symbolBeingDefined">The outer symbol being defined (to avoid recursive ids).</param>
         /// <param name="subTermAction">The action to apply to syntactic sub terms of this type.</param>
-        public static void BuildTypeId(this ITypeSymbol type, Context cx, TextWriter trapFile, Action<Context, TextWriter, ITypeSymbol> subTermAction)
+        public static void BuildTypeId(this ITypeSymbol type, Context cx, TextWriter trapFile, bool prefix, ISymbol symbolBeingDefined, Action<Context, TextWriter, ITypeSymbol, ISymbol> subTermAction)
         {
-            if (type.SpecialType != SpecialType.None)
+            if (type.SpecialType != SpecialType.None && !(type is INamedTypeSymbol n && n.IsGenericType))
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs]: Things like `IEnumerable`.

File: csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs [hunk 3]
```diff
@@ -122,6 +122,24 @@ public static bool ContainsTypeParameters(this ITypeSymbol type, Context cx, ISy
...
+        /// </summary>
+        /// <param name="type">The type to write.</param>
+        /// <param name="cx">The extraction context.</param>
+        /// <param name="trapFile">The trap file to write to.</param>
+        /// <param name="symbolBeingDefined">The outer symbol being defined (to avoid recursive ids).</param>
+        public static void BuildNestedTypeId(this ITypeSymbol type, Context cx, TextWriter trapFile, ISymbol symbolBeingDefined)
+        {
+            void WriteType(Context cx, TextWriter trapFile, ITypeSymbol symbol, ISymbol symbolBeingDefined)
+            {
+                symbol.BuildTypeId(cx, trapFile, false, symbolBeingDefined, WriteType);
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs:138]: It's a recursive lambda that needs a given type signature. So the only way I can see to name it is to use a local/static function.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.cs [hunk 4]
```diff
@@ -199,29 +199,26 @@ public override void WriteId(TextWriter trapFile)
...
         {
-            if (type.ContainsTypeParameters(cx, method))
-                type.BuildTypeId(cx, trapFile, (cx0, tb0, type0) => AddSignatureTypeToId(cx, tb0, method, type0));
-            else
-                trapFile.WriteSubId(Type.Create(cx, type));
+            type.BuildTypeId(cx, trapFile, false, symbolBeingDefined, (cx0, tb0, type0, g) => AddSignatureTypeToId(cx, tb0, method, type0, g));
         }
 
-        protected static void AddParametersToId(Context cx, TextWriter trapFile, IMethodSymbol method)
+        protected static void AddParametersToId(Context cx, TextWriter trapFile, IMethodSymbol method, ISymbol symbolBeingDefined)
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.cs:207]: That's right, so I've removed the parameter.

File: csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/SymbolExtensions.cs:138]: Ah, I didn't see that it called itself recursively.

File: csharp/ql/test/library-tests/csharp8/NullableRefTypes.expected [hunk 5]
```diff
@@ -241,9 +241,9 @@ expressionTypes
 | NullableRefTypes.cs:17:29:17:32 | null | null |
 | NullableRefTypes.cs:18:31:18:34 | this access | MyClass! |
 | NullableRefTypes.cs:19:33:19:36 | this access | MyClass! |
-| NullableRefTypes.cs:26:44:26:53 | throw ... | MyClass![]! |
+| NullableRefTypes.cs:26:44:26:53 | throw ... | MyClass?[]! |
 | NullableRefTypes.cs:26:50:26:53 | null | null |
-| NullableRefTypes.cs:27:44:27:53 | throw ... | MyClass![]! |
+| NullableRefTypes.cs:27:44:27:53 | throw ... | MyClass?[]! |
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/ql/test/library-tests/csharp8/NullableRefTypes.expected:246]: This was due to a bug in the extractor that I have now fixed.

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 6]
```diff
@@ -200,7 +215,17 @@ public void AddFreshLabel(IEntity entity)
...
+        {
+            IEqualityComparer<ISymbol> comparer = SymbolEqualityComparer.IncludeNullability;
+
+            bool IEqualityComparer<object>.Equals(object x, object y) => comparer.Equals((ISymbol)x, (ISymbol)y);
+
+            int IEqualityComparer<object>.GetHashCode(object obj) => comparer.GetHashCode((ISymbol)obj);
+        }
+
+        readonly IDictionary<object, ICachedEntity> objectEntityCache = new Dictionary<object, ICachedEntity>();
+        readonly IDictionary<object, ICachedEntity> symbolEntityCache = new Dictionary<object, ICachedEntity>(10000, new SymbolComparer());
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs:228]: We should be able to change `object` to `ISymbol` here and in `SymbolComparer`, right?

Comment (calumgrant) [reply csharp/extractor/Semmle.Extraction/Context.cs]: I originally had some difficulty converting an `IDictionary<ISymbol, ICachedEntity>` to `IDictionary<object, ICachedEntity>` but managed to work around it in the end. The result is better, so thanks!

Comment (hvitved) [reply csharp/extractor/Semmle.Extraction/Context.cs]: 🎉

**Pull request merged**